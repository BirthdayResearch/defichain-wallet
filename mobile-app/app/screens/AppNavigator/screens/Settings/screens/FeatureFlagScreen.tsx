import { ThemedScrollView, ThemedText, ThemedView } from '@components/themed'
import { Switch } from '@components/index'
import { tailwind } from '@tailwind'
import { translate } from '@translations'
import { useEffect, useState } from 'react'
import { View } from 'react-native'
import { FeatureFlag, FEATURE_FLAG_ID } from '@shared-types/website'
import { useFeatureFlagContext } from '@contexts/FeatureFlagContext'
import { WalletAlert } from '@components/WalletAlert'

export interface BetaFeaturesI extends FeatureFlag {
  value: boolean
}

export function FeatureFlagScreen (): JSX.Element {
  const { featureFlags, enabledFeatures, updateEnabledFeatures } = useFeatureFlagContext()
  const [betaFeatures, setBetaFeatures] = useState<BetaFeaturesI []>([])

  const getBetaFeature = (flags: FEATURE_FLAG_ID[]): BetaFeaturesI[] => {
    return featureFlags.reduce((features: BetaFeaturesI[], item: FeatureFlag) => {
      if (item.stage === 'beta') {
        features.push({
          ...item,
          value: flags.includes(item.id)
        })
      }
      return features
    }, [])
  }

  useEffect(() => {
    setBetaFeatures(getBetaFeature(enabledFeatures))
  }, [])

  const onFeatureChange = async (feature: FeatureFlag, value: boolean): Promise<void> => {
    const flags: FEATURE_FLAG_ID[] = value ? [...enabledFeatures, feature.id] : enabledFeatures.filter(e => e !== feature.id)
    if (value) {
      WalletAlert({
        title: translate('screens/FeatureFlagScreen', 'Enable {{feature}} (Beta)', { feature: translate('screens/Settings', feature.name) }),
        message: translate(
          'screens/FeatureFlagScreen', 'This feature is still in Beta, upon activation you will be expose to some risks. Do you want to continue?'),
        buttons: [
          {
            text: translate('screens/FeatureFlagScreen', 'Cancel'),
            style: 'cancel'
          },
          {
            text: translate('screens/FeatureFlagScreen', 'Continue'),
            style: 'destructive',
            onPress: async () => {
              setBetaFeatures(getBetaFeature(flags))
              await updateEnabledFeatures(flags)
            }
          }
        ]
      })
    } else {
      setBetaFeatures(getBetaFeature(flags))
      await updateEnabledFeatures(flags)
    }
  }

  return (
    <ThemedScrollView testID='features_flag_screen'>
      <View style={tailwind('flex-1 p-4 pt-6')}>
        <ThemedText
          style={tailwind('text-base font-semibold')}
        >
          {translate('screens/FeatureFlagScreen', 'Beta Features')}
        </ThemedText>

        <ThemedText
          dark={tailwind('text-gray-400')}
          light={tailwind('text-gray-500')}
          style={tailwind('text-sm font-normal')}
        >
          {translate('screens/FeatureFlagScreen', 'Beta feature(s) of Light Wallet represent features undergoing final testing before its official release. Experimentation of the feature(s) are encouraged, but caution is advised when using your assets.')}
        </ThemedText>
      </View>
      {betaFeatures.map((item: BetaFeaturesI) => (
        <FeatureFlagItem
          key={item.id}
          item={item}
          onChange={onFeatureChange}
        />
      ))}
    </ThemedScrollView>
  )
}

interface FeatureFlagItemProps {
  item: BetaFeaturesI
  onChange: (feature: FeatureFlag, value: boolean) => void
}

export function FeatureFlagItem ({ item, onChange }: FeatureFlagItemProps): JSX.Element {
  return (
    <View testID={`feature_${item.id}_row`}>
      <ThemedView
        dark={tailwind('bg-gray-800 border-b border-gray-700')}
        light={tailwind('bg-white border-b border-gray-200')}
        style={tailwind('flex flex-row p-4 pr-2 items-center justify-between')}
      >
        <ThemedText
          dark={tailwind('text-white text-opacity-90')}
          light={tailwind('text-black')}
          style={tailwind('font-medium')}
        >
          {translate('screens/FeatureFlagScreen', item.name)}
        </ThemedText>

        <View style={tailwind('flex-row items-center')}>
          <Switch
            onValueChange={(v: boolean) => {
            onChange(item, v)
          }}
            testID={`feature_${item.id}_switch`}
            value={item.value}
          />
        </View>
      </ThemedView>
      <ThemedText
        dark={tailwind('text-gray-400')}
        light={tailwind('text-gray-500')}
        style={tailwind('px-4 py-2 mb-2 text-sm font-normal')}
      >
        {translate('screens/FeatureFlagScreen', item.description)}
      </ThemedText>
    </View>
  )
}
