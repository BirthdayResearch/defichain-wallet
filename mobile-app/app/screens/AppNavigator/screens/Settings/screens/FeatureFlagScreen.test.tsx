import { render } from '@testing-library/react-native'
import { FeatureFlagScreen, FeatureFlagItem, BetaFeaturesI } from './FeatureFlagScreen'
import { EnvironmentNetwork } from '@environment'

jest.mock('@shared-contexts/ThemeProvider')
jest.mock('@contexts/FeatureFlagContext')

describe('feature flag screen', () => {
  it('should render FeatureFlagItem', async () => {
    const feature: BetaFeaturesI = {
      id: 'loan',
      name: 'Decentralized Loans',
      stage: 'beta',
      version: '>=0.12.0',
      description: 'Browse loan tokens provided by DeFiChain',
      value: true,
      networks: [EnvironmentNetwork.LocalPlayground, EnvironmentNetwork.RemotePlayground],
      platforms: ['ios', 'android', 'web']
    }
    const rendered = render(<FeatureFlagItem item={feature} onChange={() => {}} />)
    expect(rendered.toJSON()).toMatchSnapshot()
  })

  it('should render FeatureFlagScreen', async () => {
    const rendered = render(<FeatureFlagScreen />)
    expect(rendered.toJSON()).toMatchSnapshot()
  })
})
