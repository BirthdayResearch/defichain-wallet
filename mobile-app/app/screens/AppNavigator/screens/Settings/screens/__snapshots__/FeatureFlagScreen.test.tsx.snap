// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feature flag screen should render FeatureFlagItem 1`] = `
<View
  testID="feature_loan_row"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 8,
          "paddingTop": 16,
        },
        Object {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderBottomColor": "rgba(229, 229, 229, 1)",
          "borderBottomWidth": 1,
          "borderLeftColor": "rgba(229, 229, 229, 1)",
          "borderRightColor": "rgba(229, 229, 229, 1)",
          "borderTopColor": "rgba(229, 229, 229, 1)",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "RegularFont",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          Array [
            Object {
              "fontFamily": "MediumFont",
              "fontWeight": "500",
            },
            Object {
              "color": "rgba(0, 0, 0, 1)",
            },
          ],
        ]
      }
    >
      Decentralized Loans
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <RCTSwitch
        accessibilityRole="switch"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onTintColor="#02B31B"
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            Object {
              "backgroundColor": "#E5E5E5",
              "borderRadius": 16,
            },
          ]
        }
        testID="feature_loan_switch"
        thumbTintColor="#FFFFFF"
        tintColor="#E5E5E5"
        value={true}
      />
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "RegularFont",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 24,
        },
        Array [
          Object {
            "fontFamily": "RegularFont",
            "fontSize": 14,
            "fontWeight": "400",
            "lineHeight": 20,
            "marginBottom": 8,
            "paddingBottom": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 8,
          },
          Object {
            "color": "rgba(115, 115, 115, 1)",
          },
        ],
      ]
    }
  >
    Browse loan tokens provided by DeFiChain
  </Text>
</View>
`;

exports[`feature flag screen should render FeatureFlagScreen 1`] = `
<RCTScrollView
  style={
    Array [
      undefined,
      Object {
        "backgroundColor": "rgba(245, 245, 245, 1)",
      },
    ]
  }
  testID="features_flag_screen"
>
  <View>
    <View
      style={
        Object {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 24,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "RegularFont",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            Array [
              Object {
                "fontFamily": "SemiBoldFont",
                "fontSize": 16,
                "fontWeight": "600",
                "lineHeight": 24,
              },
              Object {
                "color": "rgba(0, 0, 0, 1)",
              },
            ],
          ]
        }
      >
        Beta Features
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontFamily": "RegularFont",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 20,
              },
              Object {
                "color": "rgba(115, 115, 115, 1)",
              },
            ],
          ]
        }
      >
        Beta feature(s) of Light Wallet represent features undergoing final testing before its official release. Experimentation of the feature(s) are encouraged, but caution is advised when using your assets.
      </Text>
    </View>
    <View
      testID="feature_loan_row"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 8,
              "paddingTop": 16,
            },
            Object {
              "backgroundColor": "rgba(255, 255, 255, 1)",
              "borderBottomColor": "rgba(229, 229, 229, 1)",
              "borderBottomWidth": 1,
              "borderLeftColor": "rgba(229, 229, 229, 1)",
              "borderRightColor": "rgba(229, 229, 229, 1)",
              "borderTopColor": "rgba(229, 229, 229, 1)",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
              Array [
                Object {
                  "fontFamily": "MediumFont",
                  "fontWeight": "500",
                },
                Object {
                  "color": "rgba(0, 0, 0, 1)",
                },
              ],
            ]
          }
        >
          Decentralized Loans
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <RCTSwitch
            accessibilityRole="switch"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            onTintColor="#02B31B"
            style={
              Array [
                Object {
                  "height": 31,
                  "width": 51,
                },
                Object {
                  "backgroundColor": "#E5E5E5",
                  "borderRadius": 16,
                },
              ]
            }
            testID="feature_loan_switch"
            thumbTintColor="#FFFFFF"
            tintColor="#E5E5E5"
            value={true}
          />
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "fontFamily": "RegularFont",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 20,
                "marginBottom": 8,
                "paddingBottom": 8,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 8,
              },
              Object {
                "color": "rgba(115, 115, 115, 1)",
              },
            ],
          ]
        }
      >
        Browse loan tokens provided by DeFiChain
      </Text>
    </View>
    <View
      testID="feature_auction_row"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 8,
              "paddingTop": 16,
            },
            Object {
              "backgroundColor": "rgba(255, 255, 255, 1)",
              "borderBottomColor": "rgba(229, 229, 229, 1)",
              "borderBottomWidth": 1,
              "borderLeftColor": "rgba(229, 229, 229, 1)",
              "borderRightColor": "rgba(229, 229, 229, 1)",
              "borderTopColor": "rgba(229, 229, 229, 1)",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
              Array [
                Object {
                  "fontFamily": "MediumFont",
                  "fontWeight": "500",
                },
                Object {
                  "color": "rgba(0, 0, 0, 1)",
                },
              ],
            ]
          }
        >
          Auction
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <RCTSwitch
            accessibilityRole="switch"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            onTintColor="#02B31B"
            style={
              Array [
                Object {
                  "height": 31,
                  "width": 51,
                },
                Object {
                  "backgroundColor": "#E5E5E5",
                  "borderRadius": 16,
                },
              ]
            }
            testID="feature_auction_switch"
            thumbTintColor="#FFFFFF"
            tintColor="#E5E5E5"
            value={false}
          />
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "fontFamily": "RegularFont",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 20,
                "marginBottom": 8,
                "paddingBottom": 8,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 8,
              },
              Object {
                "color": "rgba(115, 115, 115, 1)",
              },
            ],
          ]
        }
      >
        Browse auctions provided by DeFiChain
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
